### DISCUSSION TITLE
Approach to testing

### DISCUSSION THREAD

**Discussion by jasonkuhrt** [nextauthjs/next-auth#1244] [Help]
We currently allow users to sign up via GitHub in our app.

We're setting up our API integration testing system wherein we will do non-mocked testing.

For these integration tests we need to seed data, including users.

Our API validates requests based on the incoming JWT bearer token.

I would like to sign users up via a seed that goes through aspects next-auth, instead of completely bypassing it.

How would we do this?

1. One idea was that we would add a credentials provider when in testing stage (aka. environment)

    1. Use Prisma to insert some user/accounts following the pattern that next-auth would have done following a sign up. We would much prefer to have an internal function I can import and use from `next-auth` here, instead of reinventing anything.
    2. Before a test runs, a token would be generated based on a user/password context. Again here we need access to some internal next-auth functions.
    3. When a test runs, it would pass the token as a bearer token header. The server would do the _exact same thing it does in production_, validating that the received token is valid etc. and our logic would lookup the user in the DB, do stuff, etc. etc.

It seems that we should be able to do the similar for any provider, including GitHub so long as we correctly mock what the provider returns (e.g. user email id)?

Any advice/feedback is most welcome.

Really curious to hear what others are doing.

**Comment by jasonkuhrt** [comment]
I guess all we really want to do is:

1. Put users into the database in the same way next-auth would for the given provider
2. Forge tokens for those users to give to our tests that subsequently validated as normal

**Comment by jasonkuhrt** [comment]
I dug into the codebase tonight and figured a few things out:

1. For credentials flow I need to configure the `jwt` callback in order to make the ID I return for user show up as `sub` in the token
2. I can import default from `next-auth/jwt` to acess `jwt.encode` which allows me in tests to forge tokens with some test secret--simple
3. So long as the tokens I am forging are synced (namely `sub` field) with users seeded before hand in the DB then I will be able to make authenticated calls to the database in tests.

The only open question I have now is, rather than manually seeding the database with data, use next-auth to perform the user sign up logic that would create both account/user data etc. so that I'm not duplicating (and thus maintaining) that logic.

**Reply by briangwaltney** [reply]
I know I'm jumping into this way late, but did you figure out how to do this? Do you have an example of what you did to forge the token? I'm doing similar integration tests and trying to figure out how to create a session programmatically on the frontend and have it verified on the backend.

Where did you use the jwt.encode and how did you sync that with the db?

**Comment by jasonkuhrt** [comment]
Ah, we should just be able to call into the adaptor methods:

1. `createUser`
2. `linkAccount`

Basically this logic: https://github.com/jasonkuhrt/next-auth/blob/b3149f24f9da72544c2c780ae2620f2c5b300994/src/server/lib/callback-handler.js#L198-L209
